walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,60 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+        if (request.getAttendees().isEmpty()) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+
+        if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return Arrays.asList();
+        }
+
+        List<Event> eventList = new ArrayList(events);
+        eventList.removeIf( e -> (Collections.disjoint(e.getAttendees(), request.getAttendees()) 
+            || e.getWhen().duration() <= 0));
+        Collections.sort(eventList, new Comparator<Event>() {
+                public int compare (Event e1, Event e2) {
+                    return TimeRange.ORDER_BY_START.compare(e1.getWhen(), e2.getWhen());
+                }
+        });
+
+        if (eventList.isEmpty()) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+        
+        List<TimeRange> availableTimes = new ArrayList();
+        if (eventList.get(0).getWhen().start() - TimeRange.START_OF_DAY > request.getDuration()) {
+            availableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY,eventList.get(0).getWhen().start(),false));
+        }
```

Comment:
It can be somewhat hard to follow the logic around this part. But so far I see no glaring issues? But I would recommend adding a lot more comments in this class so if someone else like Derrill looks it over it'll be quicker for him to see what's going on (Incidentally, a scheduler-related question happens to be one of my go-to interview questions, so I already had some familiarity with this solution).

Reply:
I added javadocs and moved some code to helper functions. I wasn't sure how to do this without adding my other commit.